### ISSUE TITLE
Instructions on properly shipping code to NPM for a module consumed by this dependency

### ISSUE THREAD

**Issue by kylemh** [martpie/next-transpile-modules#121]
**Is your feature request related to a problem? Please describe.**
I'm not sure how I need to build my packages to ensure that they work properly with this dependency. I would love if my output targets and the needed keys in `package.json` were outlined, cuz modules are hard.

**Describe the solution you'd like**
I _THINK_ that I need to build and ship ESM via the `modules` key in `package.json` for this dependency to work properly, but I'm not certain if I could instead just ship source code and only have a `files` key.

**Comment by michaelyuen** [comment]
Modules _are_ hard.

In this case, we only need to consider two types, so the problem is somewhat easier. Those two types are cjs (commonjs) and esm (ecmascript / javascript modules). I have found [this 2ality post](https://2ality.com/2017/04/setting-up-multi-platform-packages.html) a great resource for wrapping my head around it.

From what you are saying, I think you're on the right track:
- ship in ESM format ✅ 
- point to the ESM code via `modules` in package.json ✅ 

`files` is important, but somewhat different. Referring to the [documentation ](https://docs.npmjs.com/files/package.json#files), it will determine which files are included in the package when installed. However, _where_ the consuming application will look based on platform goes back to `main` vs `module` (and `browser`).

>  I'm not certain if I could instead just ship source code and only have a files key.

I wouldn't advise it. I can't say I know for sure, but I don't think it would work at all. I would suggest publishing both cjs and esm formats.

**Comment by kylemh** [comment]
Thanks for the answer! Assuming I have the next bundle analyzer integrated, my guess for confirming this is correct would be to adjust my `.browserslistrc` to ensure that the bundle adjusts in size per build if I only change that from... say... IE11 to latest Chrome.

Is there an easier way?

**Comment by michaelyuen** [comment]
You're welcome! Let me preface by saying I don't know nextjs in-depth, so my comments in this regard are more generalized. 

I played around a bit with different browserslist configurations and I didn't see any meaningful differences. That could simply be because I did something wrong. 🤷‍♂️ 

The primary goal for consuming ESM is that they can be tree-shaken. So as you alluded to, the meaningful difference would be the size of your application's bundle(s). If your app is consuming cjs, I would expect it to be functionally identical, but presumably have larger bundle(s).

In my testing, I could see a clear difference in bundle size when cjs was used versus esm. The package that led me to this repo is `react-syntax-highlighter`. Due to its unique nature, it requires a line like `import { docco } from 'react-syntax-highlighter/dist/esm/styles/hljs';` so whether or not I'm consuming cjs or esm is explicit.

To summarize - the difference in bundle size should be a good general indicator. Bundle analyzer can be great to visually verify that certain dependencies are not included. Finally, you _can_ dig into the code of the built bundles and verify that the code is tree-shaken, but I wouldn't call that easy. If you're authoring your own library it may not be so bad as you can contrive very specific things to use for identifying if certain code is bundled vs tree-shaken.

**Comment by martpie** [comment]
> I'm not sure how I need to build my packages to ensure that they work properly with this dependency.

I'd need more information to understand what are your requirements, but more generally, this is something I would advice against (especially publishing, go for a monorepo instead if you can). This plugin is a Next.js hack, so use as much of `commonjs` as possible.

Now, let's assume you absolutely want to do that (really, don't), this plugin will transpile any code with the same Babel/Webpack configuration as Next.js.

It means, you can basically go full ESM/ES6/7+ for your plugins without an issue, as those are covered by Next.js Babel configuration.

If your question is "how to ship ESM packages at all?", I think this is out of scope of this plugin, and you will find plenty of tutorials out there to explain you how to ship a package on `npm`.

If you want some inspiration, you can check my other packages on my GitHub profile (making use of TSC mostly), hope that'll help :)

**Comment by kylemh** [comment]
My question is mainly how to get my packages into a shape where they can be properly consumed by this library.

**Comment by martpie** [comment]
I don't think it's in the scope of this project to tell people how to structure their `package.json` files to publish packages: publishing packages is universal and works the same everywhere: the `main` field is the entrypoint of the package, the optional `files` key list the files that gets included in your tar file, and that's about it. Nothing specific with the scope of `next-transpile-modules`, this is universal.

Example:

https://github.com/martpie/next-transpile-modules/blob/e824eba5f12af17de40dab49255bf100c9e3f3a8/package.json#L4

**Comment by michaelyuen** [comment]
I am curious and would like to ask if you could expand on this further:

> This plugin is a Next.js hack, so use as much of commonjs as possible.

Is this comment more directed at the "Next.js hack" part? Because transpiling and tree-shaking ESM is meaningful, so maybe I'm misunderstanding?

**Comment by kylemh** [comment]
I guess I was just curious if we need to ship CJS or ESM for this library to be able to properly target the dependency for transpilation.

**Comment by martpie** [comment]
> Is this comment more directed at the "Next.js hack" part? Because transpiling and tree-shaking ESM is meaningful, so maybe I'm misunderstanding?

By hack, I mean this plugin modifies the Webpack configuration provided by Next.js, and we do weird operations on it to disallow some limitations that come with it (like `node_modules` being ignored by Babel, etc). It is a functional, non-clean solution, to a real problem. This is what I meant by "hack" ;)

**Comment by martpie** [comment]
> I guess I was just curious if we need to ship CJS or ESM for this library to be able to properly target the dependency for transpilation.

If you ship CJS, you don't need this plugin ;)

**Comment by kylemh** [comment]
I have different apps that have different browserslist targets. I have one library used in all of the apps that I DONT want to ship CJS such that each app can transpile to their needs. Does that context help? Thanks for bearing with me.

**Comment by michaelyuen** [comment]
> I DONT want to ship CJS such that each app can transpile to their needs

Makes sense. Then you'll want to ship ESM.

> I'm not certain how the dependencies themselves should ship such that everything transpiles properly

They should ship in ESM format - import & export.

**Comment by kylemh** [comment]
I'm going to be acting on your advice for sure! Just tryna convey the thought to @martpie in case there's room for a PR to be made in this repo on this matter, but no worries if not!

**Comment by martpie** [comment]
I did not think about the browserlist thing, I usually use the default polyfills + import some manually when needed. I'll need to think a bit about it.

**Comment by martpie** [comment]
@kylemh can you show me your Babel config?

**Comment by kylemh** [comment]
```
{
  "presets": ["next/babel"],
  "env": {
    "development": {
      "plugins": [["styled-components", { "displayName": true, "ssr": true }], "polished"]
    },
    "production": {
      "plugins": [["styled-components", { "displayName": false, "ssr": true }], "polished"]
    }
  }
}
```

Next.js's babel config: https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts

**Comment by martpie** [comment]
In the FAQ https://github.com/martpie/next-transpile-modules#i-have-trouble-with-transpilation-and-my-custom-babelrc

**Comment by kylemh** [comment]
Are you saying that next.js app's can load the babel config from `babel.config.js` too, not just `.babelrc`?
